8283091: Support type conversion between different data sizes in SLP

fg1417 opened the pull request:
After JDK-8275317, C2's SLP vectorizer has supported type conversion between the same data size. We can also support conversions between different data sizes like:
int <-> double
float <-> long
int <-> long
float <-> double

A typical test case:

int[] a;
double[] b;
for (int i = start; i < limit; i++) {
    b[i] = (double) a[i];
}

Our expected OptoAssembly code for one iteration is like below:

add R12, R2, R11, LShiftL #2
vector_load   V16,[R12, #16]
vectorcast_i2d  V16, V16  # convert I to D vector
add R11, R1, R11, LShiftL #3	# ptr
add R13, R11, #16	# ptr
vector_store [R13], V16

To enable the vectorization, the patch solves the following problems in the SLP.

There are three main operations in the case above, LoadI, ConvI2D and StoreD. Assuming that the vector length is 128 bits, how many scalar nodes should be packed together to a vector? If we decide it separately for each operation node, like what we did before the patch in SuperWord::combine_packs(), a 128-bit vector will support 4 LoadI or 2 ConvI2D or 2 StoreD nodes. However, if we put these packed nodes in a vector node sequence, like loading 4 elements to a vector, then typecasting 2 elements and lastly storing these 2 elements, they become invalid. As a result, we should look through the whole def-use chain
and then pick up the minimum of these element sizes, like function SuperWord::max_vector_size_in_ud_chain() do in the superword.cpp. In this case, we pack 2 LoadI, 2 ConvI2D and 2 StoreD nodes, and then generate valid vector node sequence, like loading 2 elements, converting the 2 elements to another type and storing the 2 elements with new type.

After this, LoadI nodes don't make full use of the whole vector and only occupy part of it. So we adapt the code in SuperWord::get_vw_bytes_special() to the situation.

In SLP, we calculate a kind of alignment as position trace for each scalar node in the whole vector. In this case, the alignments for 2 LoadI nodes are 0, 4 while the alignment for 2 ConvI2D nodes are 0, 8. Sometimes, 4 for LoadI and 8 for ConvI2D work the same, both of which mark that this node is the second node in the whole vector, while the difference between 4 and 8 are just because of their own data sizes. In this situation, we should try to remove the impact caused by different data size in SLP. For example, in the stage of SuperWord::extend_packlist(), while determining if it's potential to pack a pair of def nodes in the function SuperWord::follow_use_defs(), we remove the side effect of different data size by transforming the target alignment from the use node. Because we believe that, assuming that the vector length is 512 bits, if the ConvI2D use nodes have alignments of 16-24 and their def nodes, LoadI, have alignments of 8-12, these two LoadI nodes should be packed as a pair as well.

Similarly, when determining if the vectorization is profitable, type conversion between different data size takes a type of one size and produces a type of another size, hence the special checks on alignment and size should be applied, like what we do in SuperWord::is_vector_use().

After solving these problems, we successfully implemented the vectorization of type conversion between different data sizes.

Here is the test data  (-XX:+UseSuperWord) on NEON:

Before the patch:
Benchmark   (length)  Mode  Cnt    Score   Error  Units
convertD2F       523  avgt   15  216.431 ± 0.131  ns/op
convertD2I       523  avgt   15  220.522 ± 0.311  ns/op
convertF2D       523  avgt   15  217.034 ± 0.292  ns/op
convertF2L       523  avgt   15  231.634 ± 1.881  ns/op
convertI2D       523  avgt   15  229.538 ± 0.095  ns/op
convertI2L       523  avgt   15  214.822 ± 0.131  ns/op
convertL2F       523  avgt   15  230.188 ± 0.217  ns/op
convertL2I       523  avgt   15  162.234 ± 0.235  ns/op

After the patch:
Benchmark   (length)  Mode  Cnt    Score    Error  Units
convertD2F       523  avgt   15  124.352 ±  1.079  ns/op
convertD2I       523  avgt   15  557.388 ±  8.166  ns/op
convertF2D       523  avgt   15  118.082 ±  4.026  ns/op
convertF2L       523  avgt   15  225.810 ± 11.180  ns/op
convertI2D       523  avgt   15  166.247 ±  0.120  ns/op
convertI2L       523  avgt   15  119.699 ±  2.925  ns/op
convertL2F       523  avgt   15  220.847 ±  0.053  ns/op
convertL2I       523  avgt   15  122.339 ±  2.738  ns/op

perf data on X86:
Before the patch:
Benchmark      (length)  Mode  Cnt    Score   Error  Units
  convertD2F       523  avgt   15  279.466 ± 0.069  ns/op
  convertD2I       523  avgt   15  551.009 ± 7.459  ns/op
  convertF2D       523  avgt   15  276.066 ± 0.117  ns/op
  convertF2L       523  avgt   15  545.108 ± 5.697  ns/op
  convertI2D       523  avgt   15  745.303 ± 0.185  ns/op
  convertI2L       523  avgt   15  260.878 ± 0.044  ns/op
  convertL2F       523  avgt   15  502.016 ± 0.172  ns/op
  convertL2I       523  avgt   15  261.654 ± 3.326  ns/op

After the patch:
Benchmark      (length)  Mode  Cnt    Score   Error  Units
  convertD2F       523  avgt   15  106.975 ± 0.045  ns/op
  convertD2I       523  avgt   15  546.866 ± 9.287  ns/op
  convertF2D       523  avgt   15   82.414 ± 0.340  ns/op
  convertF2L       523  avgt   15  542.235 ± 2.785  ns/op
  convertI2D       523  avgt   15   92.966 ± 1.400  ns/op
  convertI2L       523  avgt   15   79.960 ± 0.528  ns/op
  convertL2F       523  avgt   15  504.712 ± 4.794  ns/op
  convertL2I       523  avgt   15  129.753 ± 0.094  ns/op

perf data on AVX512:
Before the patch:
Benchmark      (length)  Mode  Cnt    Score   Error  Units
  convertD2F       523  avgt   15  282.984 ± 4.022  ns/op
  convertD2I       523  avgt   15  543.080 ± 3.873  ns/op
  convertF2D       523  avgt   15  273.950 ± 0.131  ns/op
  convertF2L       523  avgt   15  539.568 ± 2.747  ns/op
  convertI2D       523  avgt   15  745.238 ± 0.069  ns/op
  convertI2L       523  avgt   15  260.935 ± 0.169  ns/op
  convertL2F       523  avgt   15  501.870 ± 0.359  ns/op
  convertL2I       523  avgt   15  257.508 ± 0.174  ns/op

After the patch:
Benchmark  (length)  Mode  Cnt    Score   Error  Units
  convertD2F       523  avgt   15   76.687 ± 0.530  ns/op
  convertD2I       523  avgt   15  545.408 ± 4.657  ns/op
  convertF2D       523  avgt   15  273.935 ± 0.099  ns/op
  convertF2L       523  avgt   15  540.534 ± 3.032  ns/op
  convertI2D       523  avgt   15  745.234 ± 0.053  ns/op
  convertI2L       523  avgt   15  260.865 ± 0.104  ns/op
  convertL2F       523  avgt   15   63.834 ± 4.777  ns/op
  convertL2I       523  avgt   15   48.183 ± 0.990  ns/op

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8283091](https://bugs.openjdk.org/browse/JDK-8283091): Support type conversion between different data sizes in SLP


### Reviewers
 * [Vladimir Kozlov](https://openjdk.org/census#kvn) (@vnkozlov - **Reviewer**) ⚠️ Review applies to [0d731bb2](https://git.openjdk.org/jdk/pull/7806/files/0d731bb2b5215ca7a28077c8e2fa913db5d67720)
 * [Sandhya Viswanathan](https://openjdk.org/census#sviswanathan) (@sviswa7 - **Reviewer**)


### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/7806/head:pull/7806` \
`$ git checkout pull/7806`

Update a local copy of the PR: \
`$ git checkout pull/7806` \
`$ git pull https://git.openjdk.org/jdk pull/7806/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 7806`

View PR using the GUI difftool: \
`$ git pr show -t 7806`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/7806.diff">https://git.openjdk.org/jdk/pull/7806.diff</a>

</details>

DamonFool commented:
Good job!

Did you forget the micro-benchmark VectorLoop.java?

fg1417 commented:
> Did you forget the micro-benchmark VectorLoop.java?

Thanks for your review. Sorry to miss it. Add it ASAP :-)

fg1417 commented:
> Did you forget the micro-benchmark VectorLoop.java?
Thanks. Done.

fg1417 commented:
The patch is to support type conversion between different data sizes like ConvL2F and ConvL2I. Anybody can help review it? Thanks :)

DamonFool commented:
Can you explain why `convertD2I` becomes much slower on NEON after this patch?
Thanks.

DamonFool commented:
There seems conflicts with the jdk-master, so please merge with the latest version.
Thanks.

fg1417 commented:
> Can you explain why `convertD2I` becomes much slower on NEON after this patch? Thanks.

Thanks for your review.

On NEON, there is no real vector instructions to do the conversion from Double to Int, which is implemented in 5 instructions https://github.com/openjdk/jdk/blob/8c1870521815a24fd12480e73450c2201542a442/src/hotspot/cpu/aarch64/aarch64_neon.ad#L512, costing more than scalar instructions, as we know that there are only two elements for ConvertD2I on 128-bit NEON machine.

DamonFool commented:
> costing more than scalar instructions, as we know that there are only two elements for VectorCastD2I on 128-bit NEON machine.

So shall we disable `vcvt2Dto2I` for NEON?

fg1417 commented:
> So shall we disable `vcvt2Dto2I` for NEON?

I'm afraid we can't. We still need to support it in VectorAPI.

vnkozlov commented:
And separate thank you for updating tests to verify correctness of this optimization!

fg1417 commented:
> Please update to latest JDK (Loom was integrated) and run performance again. Also include % of changes.
> 
> I have the same concern as @DamonFool about regression when vectorizing some conversions. May be we should have additional `Matcher` property we could consult when trying to **auto-vectorize**. I understand that we need `vcvt2Dto2I` when VectorAPI specifically asking to generate it but we should not enforce auto-generation.

@vnkozlov thanks for your review and kind suggestion! I'll update the patch to resolve the potential performance regression.

vnkozlov commented:
@fg1417  I don't see new update in this PR.  Please also show performance numbers with new changes

fg1417 commented:
> @fg1417 I don't see new update in this PR. Please also show performance numbers with new changes

Here is the perf uplift data (ns/op) on different machines for the latest patch.

  NEON          perf change (ns/op)
convertB2D      not supported
convertB2F      -45.55%
convertB2L      not supported
convertD2B      not supported
convertD2F      -42.32%
convertD2I      not supported (VectorAPI supported)
convertD2S      not supported
convertF2B      -42.95%
convertF2D      -45.28%
convertF2L      -5.78%
convertF2S      -51.30%
convertI2D      -27.82%
convertI2L      -44.54%
convertL2B      not supported
convertL2F      not supported (VectorAPI supported)
convertL2I      -28.58%
convertL2S      not supported
convertS2D      not supported
convertS2F      -53.37%
convertS2L      not supported

  SVE           perf change (ns/op)
convertB2D      -36.15%
convertB2F      -63.48%
convertB2L      -32.48%
convertD2B      0.02%
convertD2F      -47.85%
convertD2I      -46.42%
convertD2S      -32.08%
convertF2B      -59.54%
convertF2D      -60.81%
convertF2L      -61.81%
convertF2S      -67.67%
convertI2D      -60.63%
convertI2L      -57.23%
convertL2B      0.04%
convertL2F      -47.21%
convertL2I      -34.49%
convertL2S      -19.57%
convertS2D      -47.20%
convertS2F      -74.86%
convertS2L      -49.00%

  X86           perf change (ns/op)
convertB2D      -64.13%
convertB2F      -79.37%
convertB2L      -70.97%
convertD2B      not supported
convertD2F      -62.69%
convertD2I      not supported
convertD2S      not supported
convertF2B      not supported
convertF2D      -68.90%
convertF2L      not supported
convertF2S      not supported
convertI2D      -87.48%
convertI2L      -69.64%
convertL2B      -3.96%
convertL2F      -0.11%
convertL2I      -49.59%
convertL2S      -24.75%
convertS2D      -84.35%
convertS2F      -86.09%
convertS2L      -70.42%

  AVX512        perf change (ns/op)
convertB2D      -78.08%
convertB2F      -86.39%
convertB2L      -79.07%
convertD2B      not supported
convertD2F      -71.86%
convertD2I      not supported
convertD2S      not supported
convertF2B      not supported
convertF2D      -78.17%
convertF2L      not supported
convertF2S      not supported
convertI2D      -90.26%
convertI2L      -79.92%
convertL2B      -70.75%
convertL2F      -86.67%
convertL2I      -80.94%
convertL2S      -71.54%
convertS2D      -90.84%
convertS2F      -83.94%
convertS2L      -80.51%

fg1417 commented:
> Do you mean it fail without #8961 or it fail always?

@vnkozlov ,after https://bugs.openjdk.java.net/browse/JDK-8286972, [the case](https://github.com/openjdk/jdk/blob/6ff2d89ea11934bb13c8a419e7bad4fd40f76759/test/hotspot/jtreg/compiler/c2/cr6340864/TestDoubleVect.java#L723) would fail without https://github.com/openjdk/jdk/pull/8961. With it, the case passes.

vnkozlov commented:
> > Do you mean it fail without #8961 or it fail always?
> 
> @vnkozlov ,after [JDK-8286972](https://bugs.openjdk.java.net/browse/JDK-8286972), [the case](https://github.com/openjdk/jdk/blob/6ff2d89ea11934bb13c8a419e7bad4fd40f76759/test/hotspot/jtreg/compiler/c2/cr6340864/TestDoubleVect.java#L723) would fail without #8961. With it, the case passes.

Got it.

fg1417 commented on src/hotspot/share/opto/vectornode.cpp:
```diff
@@ -248,14 +248,6 @@ int VectorNode::opcode(int sopc, BasicType bt) {
     return Op_StoreVector;
   case Op_MulAddS2I:
     return Op_MulAddVS2VI;
-  case Op_ConvI2F:
-    return Op_VectorCastI2X;
-  case Op_ConvL2D:
-    return Op_VectorCastL2X;
-  case Op_ConvF2I:
-    return Op_VectorCastF2X;
-  case Op_ConvD2L:
-    return Op_VectorCastD2X;
```
@vnkozlov @sviswa7 , I removed these lines here, because I call another api to generate vector opcodes and vector nodes for conversion nodes [VectorCastNode::opcode()](https://github.com/openjdk/jdk/blob/ba9ee8cb286268f1d6a2820508334aaaf3131e15/src/hotspot/share/opto/vectornode.cpp#L1203), which is more precise than this one and unified with Vector API. Because converting short to double and converting int to double would use the same opcode `ConvI2D` for scalar nodes, we can't differentiate between them just based on the opcode of the current node. I adjusted the callers `VectorCastNode::implemented()` and `SuperWord::output()` to the removing.

fg1417 replied:
> May be we should have assert here to make sure that in all places we call `VectorCastNode::opcode()` for `Conv*` nodes:
> 
> ```
>  default:
>     assert(!VectorNode::is_convert_opcode(sopc), "Convert node %s should be processed by VectorCastNode::opcode()", NodeClassNames[sopc]);
>     return 0; // Unimplemented
> ```

Make sense to me. Thanks!

fg1417 commented:
> // And exclude cases which are not profitable to auto-vectorize.

Done.

> Put it into a separate function because this code pattern is used 2 times.

Done.

> May be we should have assert here to make sure that in all places we call `VectorCastNode::opcode()` for `Conv*` nodes

Done.

Fixed the comments above and rebased to the latest JDK.
Thanks.

vnkozlov commented:
> > @fg1417 Please do not rebase or force-push to an active PR as it invalidates existing review comments. All changes will be squashed into a single commit automatically when integrating. See [OpenJDK Developers’ Guide](https://openjdk.java.net/guide/#working-with-pull-requests) for more information.
> 
> May I ask if I did anything wrong? I just rebased the master, resolved conflict and pushed a new commit as it guides... and did not do any force-push... Why I got the notification this time?

Which `git` instruction you used? It is recommended to use `git merge master` in the PR's branch after `master` branch update (**8. Merge the latest changes**):
```
Avoid rebasing changes, and prefer merging instead.
```

vnkozlov commented:
But only webrev was affected so it is fine.

vnkozlov commented on src/hotspot/share/opto/matcher.hpp:
```diff
@@ -325,6 +325,10 @@ class Matcher : public PhaseTransform {
 
+  // Identify extra cases that we might want to vectorize automatically
+  // And exclude cases which are not profitable to auto-vectorize.
+  static const bool match_rule_supported_vectorization(int opcode, int vlen, BasicType bt);
```
May be we should rename it to `match_rule_supported_superword`. That is how we call auto-vectorizer in C2.

fg1417 replied:
> May be we should rename it to `match_rule_supported_superword`. That is how we call auto-vectorizer in C2.

Done. 

And update after https://github.com/openjdk/jdk/pull/8877. Thanks.

vnkozlov reviewed (commented):
Latest changes look good. I have one comment about method name.

These changes are related to #8877 which also limit vectors in auto-vectorizer. #8877 has priority since it fixed that issue.

Also JDK 19 fork is coming in 3 day: https://openjdk.java.net/projects/jdk/19/

I would advice you to push these changes into JDK 20 after fork.

I will start my testing and do approval based on results.

vnkozlov reviewed (commented):
Good.

I submitted testing.

vnkozlov commented:
Please consider adding IR framework test to make sure expected vector nodes are generated.

vnkozlov reviewed (approved):
Testing results are good.

fg1417 commented:
> Please consider adding IR framework test to make sure expected vector nodes are generated.

Added an IR framework testcase and updated to the latest JDK. 

Thanks for your review @vnkozlov .

vnkozlov commented:
Thank you for test. @sviswa7 can you review latest changes to have 2 reviews?

sviswa7 commented:
Looks good to me.

fg1417 commented:
Thanks for your review @sviswa7 @vnkozlov .

Can I integrate it now?

vnkozlov commented:
> Thanks for your review @sviswa7 @vnkozlov .
> 
> Can I integrate it now?

Yes

fg1417 commented:
/integrate

pfustc commented:
/sponsor

TobiHartmann commented:
This change introduced a regression, see [JDK-8290910](https://bugs.openjdk.org/browse/JDK-8290910).
